feat: Mark builders to be const

Matt3o12 opened the pull request:
Hello thank you for your work. 

I have made changes so that all builders are now const. They should still be 100% backwards compadible. I had to introduce a constructor, new, as Default cannot be const just yet.

This allows me to define a const builder at the top and use it like this:

```
    const BACKOFF_BUILDER: ExponentialBuilder = ExponentialBuilder::new().with_min_delay(Duration::from_millis(1));

    fn test() {
        my_fn().retry(BACKOFF_BUILDER).call();
    }
```


All changes:

- Refactored all methods inside FibonacciBuilder, ExponentialBuilder, and ConstantBuilder to be const.
- Added a new const `new` constructor for each builder which uses the values from default()
- Updated the default method to call `new()`.
- Added unit tests for the const builders to ensure they have the correct values.

Xuanwo commented:
Hi @Matt3o12, thank you for your work first! I believe we can remove that `debug_assert`. Would you like to update the documentation for `with_factor` to emphasize the importance of ensuring that `factor > 1`?

Xuanwo commented:
Hi @Matt3o12, are you interested in moving this PR forward?

Matt3o12 commented:
Hey, sorry about the late reply, I have forgotten about it. I have now removed the `debug_assert` and updated the docs. Please let me know if you need me to change anything or to roll up the commits.

Xuanwo reviewed (approved):
Thank you @Matt3o12 for working on this!

Xuanwo commented:
> Edit: it seems you have to manually rerun the windows job or maybe empty the cache.

It's not related to this PR. embassy-time seems not work correctly on windows.
